aten/src/ATen/native/transformers/cuda/sdp_utils.h

```diff
@@ -246,21 +248,77 @@ inline bool check_tensor_shapes(sdp_params params, bool debug) {
+        if (debug) {
+          TORCH_WARN(
+            "Both fused kernels requires query, key and value to have broadcastable ",
+            param_name,
+            "Query ",
+            param_name,
+            q_size,
+            ", Key ",
+            param_name,
+            k_size,
+            ", Value ",
+            param_name,
+            v_size,
+            " instead.");
+        }
+        return false;
+  }
+  return true;
+}
+
+inline bool check_batch_size_and_num_heads(sdp_params params, bool debug) {
   // This is expected to be called after check_tensor_shapes ensuring that the size()
   // calls won't error since the inputs are all 4 dimensional
-  bool same_batch_size = params.query.size(0) == params.key.size(0) &&
-      params.query.size(0) == params.value.size(0);
-  // We pass through for NestedTensors since this is checked in a later filter
-  bool same_num_heads = params.query.is_nested()
-      ? true
-      : params.query.size(1) == params.key.size(1) &&
-          params.query.size(1) == params.value.size(1);
+  auto q_batch_size = params.query.size(0);
+  auto k_batch_size = params.key.size(0);
+  auto v_batch_size = params.value.size(0);
+
+  auto q_num_heads = params.query.size(1);
+  auto k_num_heads = params.key.size(1);
+  auto v_num_heads = params.value.size(1);
+
+  // Flip because check_for_nested_inputs returns false if there exists a nested input
+  bool has_nested_input = !check_for_nested_inputs(params, false);
+  bool same_batch_size = q_batch_size == k_batch_size && q_batch_size == v_batch_size;
+  bool same_num_heads = q_num_heads == k_num_heads && q_num_heads == v_num_heads;
+
+  if (has_nested_input) {
+    bool broadcastable_batch_size = true, broadcastable_num_heads = true;
+    if (!same_batch_size) {
+      // try to broadcast batchsize
+      broadcastable_batch_size = try_broadcast_param_size(q_batch_size,
+                                                          k_batch_size,
+                                                          v_batch_size,
+                                                          "batch size ",
+                                                          debug);
+    }
+
+    if (!same_num_heads) {
+      broadcastable_num_heads = try_broadcast_param_size(q_num_heads,
+                                                         k_num_heads,
+                                                         v_num_heads,
+                                                         "num heads ",
+                                                         debug);
+    }
+    return broadcastable_batch_size && broadcastable_num_heads;
+  }
 
   if (!(same_batch_size && same_num_heads)) {
     if (debug) {
       TORCH_WARN(
-        "Both fused kernels requires query, key and value to have the same batch_size and num_heads. Query.sizes(): ",
+        "For dense inputs, both fused kernels require query, key and value to have the same batch_size and num_heads. ",
```

Comment:
Might be worth saying, try unsqueeze and expand_to but I think that if they are savy enough to turn off the math fallback they will probably know that.